webviz_config/webviz_factory_registry.py

```diff
@@ -0,0 +1,84 @@
+from typing import Dict, TypeVar, Type, Optional, Any
+
+from .webviz_instance_info import WebvizInstanceInfo
+
+# pylint: disable=invalid-name
+T = TypeVar("T")
+
+
+class WebvizFactoryRegistry:
+    """Global registry for factories that allows the actual factory instances
+    to be shared between plugins. Also facilitates storage of optional factory
+    settings that are read from the YAML config file for later consumption when
+    an actual factory gets created and added to the registry.
+    The registry is exposed globally through WEBVIZ_FACTORY_REGISTRY below.
+    This is also the reason for the two-stage initialization approach. Note that
+    the registry instance is useless until the initialize() method has been called.
+    """
+
+    def __init__(self) -> None:
+        self._is_initialized: bool = False
+        self._app_instance_info: Optional[WebvizInstanceInfo] = None
+        self._factory_settings_dict: Dict[str, Any] = {}
+        self._factories: Dict[Type, Any] = {}
+
+    def initialize(
+        self,
+        app_instance_info: WebvizInstanceInfo,
+        factory_settings_dict: Optional[Dict[str, Any]],
+    ) -> None:
+        """Does the actual initialization of the object instance.
+        This function will be called as part of the webviz_app.py / jinja2 template.
+        """
+        if self._is_initialized:
+            raise ValueError("Registry already initialized")
+
+        if not isinstance(app_instance_info, WebvizInstanceInfo):
+            raise TypeError("app_instance_info must be of type WebvizInstanceInfo")
+        self._app_instance_info = app_instance_info
+
+        if factory_settings_dict:
+            if not isinstance(factory_settings_dict, dict):
+                raise TypeError("factory_settings_dict must be of type dict")
+            self._factory_settings_dict = factory_settings_dict
+
+        self._is_initialized = True
+
+    def set_factory(self, factory_class: Type[T], factory_obj: T) -> None:
+        if not self._is_initialized:
+            raise ValueError("Illegal access, factory registry is not initialized")
+
+        if not isinstance(factory_obj, factory_class):
+            raise TypeError("The type of the factory does not match factory_class")
+
+        self._factories[factory_class] = factory_obj
+
+    def get_factory(self, factory_class: Type[T]) -> Optional[T]:
+        if not self._is_initialized:
+            raise ValueError("Illegal access, factory registry is not initialized")
```

Comment:
`RuntimeError` instead, or maybe a new custom error is worth it since "Factory registry is not initalized" can happen multiple places?

https://docs.python.org/3/tutorial/errors.html#user-defined-exceptions

Reply:
Maybe `RuntimeError` is best suited?
It doesn't seem worth it to introduce a custom exception since the only time this error should occur is if `WebvizFactoryRegistry` is used without being initialized. Normal usage of the registry is via `WEBVIZ_FACTORY_REGISTRY` and in that case our own jinja2 templates are resposnible for initialization.